two-qubit decomposition into sqrt(iSWAP) (aka SQiSW)

ajavadia opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x ] I have added the tests to cover my changes.
- [ x] I have updated the documentation accordingly.
- [ x] I have read the CONTRIBUTING document.
-->
 
### Summary
Adds the sqrt(iSWAP) as well as a method to decompose into it that uses 2 applications for ~79% of Haar-random unitaries and 3 applications for the rest.

Based on https://arxiv.org/pdf/2105.06074v3.pdf.  This does not exactly follow the paper but is inspired by that result. The algorithm given in the paper seems to have some bugs so this is a different implementation.

Cryoris commented:
With our gate naming scheme, should this be called `SiSwapGate`? Since we started to use the single letter `S` to indicate square roots in the `SXGate` 🙂

ajavadia commented:
> With our gate naming scheme, should this be called SiSwapGate? Since we started to use the single letter S to indicate square roots in the SXGate 🙂

I started the branch with that naming scheme but quickly got annoyed at all the up and down of the camel case :) 
Still I don't particularly mind either way. I'd not that SZ is just called S at the moment because that's more common

ajavadia commented:
Also @Cryoris would you like to review this? :)

Cryoris commented on test/python/synthesis/test_sqisw_synthesis.py:
```diff
@@ -0,0 +1,68 @@
+
+
+@ddt
+class TestSQiSWSynth(QiskitTestCase):
```
It would be nice to add a test for each of the different regions (green, blue, brown and red) in the paper so we're sure each of the cases is triggered 🙂

ajavadia replied:
i added a bunch of tests

Cryoris commented on qiskit/circuit/library/standard_gates/sqisw.py:
```diff
@@ -0,0 +1,77 @@
+        from .rxx import RXXGate
+        from .ryy import RYYGate
+
+        q = QuantumRegister(2, "q")
```
Could you add this decomposition as an entry in the equivalence library? 🙂 Also, are there other convenient equivalences that we'd like to add or do we just want to stick with the unitary decomposition for now?

Cryoris reviewed (commented):
The decomposition looks mostly good! I only encountered one case where the final decomposition doesn't actually include the `SQiSWGate` but just an `Gate` with the same name, I think due to taking the inverse of the decomposition. We could e.g. fix that by adding a `SQiSWdgGate` but maybe there's also another solution. Here's a reproducing example: 

```python
import numpy as np
from qiskit.circuit import QuantumCircuit
from qiskit.circuit.library import SQiSWGate
from qiskit.synthesis.two_qubit import SQiSWDecomposer
from qiskit.quantum_info import Operator

circuit = QuantumCircuit(2)
circuit.rxx(0.2, 0, 1)
circuit.ryy(0.2, 0, 1)
circuit.rzz(0.2, 0, 1)

op = Operator(circuit).data

decomp = SQiSWDecomposer(euler_basis=["u"])
out = decomp(op)

print(out.draw())

for inst in out.data:
    if len(inst.qubits) == 2:
        print(type(inst.operation))
```
which prints
```
global phase: 2.6038
     ┌─────────────────────┐┌────────┐ ┌─────────────────────────┐  ┌────────┐┌────────────────────────────┐┌────────┐┌────────────────────────────┐
q_0: ┤ U(0,2.2455,-2.2455) ├┤0       ├─┤ U(2.9419,1.2433,2.4222) ├──┤0       ├┤ U(0.34939,2.4962,-0.64543) ├┤0       ├┤ U(2.9419,-0.71935,-1.8983) ├
     └──┬────────────────┬─┘│  sqisw │┌┴─────────────────────────┴─┐│  sqisw │└──┬─────────────────────┬───┘│  sqisw │├────────────────────────────┤
q_1: ───┤ U(-π,π/2,-π/2) ├──┤1       ├┤ U(-2.2067,-2.4822,0.54952) ├┤1       ├───┤ U(-1.0943,-π/2,π/2) ├────┤1       ├┤ U(-0.93488,2.5921,-2.4822) ├
        └────────────────┘  └────────┘└────────────────────────────┘└────────┘   └─────────────────────┘    └────────┘└────────────────────────────┘
<class 'qiskit.circuit.gate.Gate'>  # <-- should be type SQiSWGate!
<class 'qiskit.circuit.gate.Gate'>
<class 'qiskit.circuit.gate.Gate'>
```

Concerning the gate itself: I think it would be more consistent to go back to the name `SiSwapGate` to be consistent with our naming scheme. We've also moved away from other aliases like "Fredkin" or "Toffoli" in favor of consistency. I'd be open to add an alias `SQiSW = SiSwapGate` but I think it would be better if the base class would follow our convention 🙂

ShellyGarion commented on qiskit/synthesis/__init__.py:
```diff
@@ -77,6 +77,13 @@
+.. autosummary::
+   :toctree: ../stubs/
+
+   SQiSWDecomposer
```
Should it be called `SQiSWDecomposer` or `SQiSWDecomposition` like `SolovayKitaevDecomposition`?

ajavadia replied:
Right now it's named like TwoQubitBasisDecomposer and XXDecomposer which are more related to this class as they all take SU(4) and decompose it over some basis.
I agree we need consistency and my intent is to do this as I move stuff over to qiskit.synthesis (over a few different PRs)

ShellyGarion commented on qiskit/synthesis/two_qubit/sqisw_decompose.py:
```diff
@@ -0,0 +1,211 @@
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.quantum_info.operators import Operator
+from qiskit.quantum_info.synthesis.two_qubit_decompose import TwoQubitWeylDecomposition
```
it's a bit confusing that this code in qiskit/sythesis whereas two_qubit_decompose is in qiskit/quantum_info/synthesis.
we should probably deprecate qiskit/quantum_info/synthesis and move all the code there to qiskit/sythesis.
(but this should not be part of this PR).

ajavadia replied:
yes that's my intent. to put the new stuff in qiskit/synthesis and then move over the old stuff too

jakelishman replied:
Doing that properly will also (hopefully) help us break a bunch of import cycles.

ShellyGarion commented on test/python/synthesis/test_sqisw_synthesis.py:
```diff
@@ -0,0 +1,68 @@
+
+@ddt
+class TestSQiSWSynth(QiskitTestCase):
+    """Test the Gray-Synth algorithm."""
```
This comment is not relevant to this test

ShellyGarion commented on qiskit/synthesis/two_qubit/sqisw_decompose.py:
```diff
@@ -0,0 +1,211 @@
+
+class SQiSWDecomposer:
+    """
+    A class for decomposing 2-qubit unitaries into at most 3 uses of the sqrt(iSWAP) gate.
```
Perhaps it's worth to explain somewhere the difference from the usual two-qubit decomposition, and when to use each of the algorithms.

ShellyGarion replied:
I would still like to check the docs, after they will be generated when the conflict is solved.

ajavadia commented:
I addressed the review comments. Also I added approximation which is a good boost to the performance of the method (details in the PR description). I also added more corner case tests.

ajavadia commented on qiskit/synthesis/two_qubit/sqisw_decompose.py:
```diff
@@ -0,0 +1,211 @@
+                for instruction in inverse_decomposition:
+                    if isinstance(instruction.operation, SQiSWGate):
+                        inverse_decomposition_with_sqisw_dg.z(0)
+                        inverse_decomposition_with_sqisw_dg.append(SQiSWGate().inverse(), [0, 1])
```
I started trying to fix this as you suggest but thinking about it I think the overhead of adding a new gate (and all that goes along with it like definitions and adding to equivalence library) is not worth it. This problem has to be fixed in Qiskit such that taking the inverse of a gate twice gives the same gate. I opened an issue on this 18 months ago https://github.com/Qiskit/qiskit-terra/issues/6879. This has become a higher priority recently and I think once it gets fixed this automatically gets fixed too. Right now it's not really a problem because if the device supports siswap gate then the generated qasm is correct.

ShellyGarion commented on qiskit/synthesis/__init__.py:
```diff
@@ -78,6 +78,13 @@
 
    SolovayKitaevDecomposition
 
+SU(4) Synthesis
```
Does anyone know what SU(4) is? Is it better to say something like "Two-qubit Unitary Synthesis"?

ajavadia replied:
```suggestion
SU(4) (2-qubit unitary) Synthesis
```

ajavadia replied:
I was sketching something out, please let me know what you think. Some of these are placeholders for things i'd like to add in the near future. 

```
qiskit.synthesis
 |--- su2
    |--- euler
    |--- nonorthogonal
 |--- su4
    |--- supercontrolled
    |--- controlled
    |--- siswap
    |--- b
 |--- unitary
    |--- aqc (approximate quantum compiling)
    |--- qsd (quantum shannon decomposition)
    |--- ccd (column-by-column)
 |--- state
    |
 |--- clifford
    |--- ag
    |--- bm
    |--- full
    |--- greedy
    |--- layers
 |--- linear
    |--- graysynth
    |--- depth_lnn
    |--- pmh
 |--- permutation  * use method/paper names here
    |--- lnn
    |--- full
 |--- cnotdihedral
    |--- full
    |--- general
    |--- two_qubits
 |--- discrete_basis
    |--- solovay-kitaev
 |--- evolution
    |--- lie_trotter
    |--- suzuki-trotter
    |--- product_formula
    |--- qdrift
 |--- povm
 ```

ShellyGarion replied:
Nice! I'll open an issue to discuss the structure of `qiskit/synthesis` (as this is not a part of this PR).

ShellyGarion replied:
See: https://github.com/Qiskit/qiskit-terra/issues/9667

ajavadia commented on qiskit/synthesis/__init__.py:
```diff
@@ -79,6 +79,13 @@
    SolovayKitaevDecomposition
    generate_basic_approximations
 
+SU(4) (2-qubit unitary) Synthesis
+===============
```
```suggestion
SU(4) (2-qubit unitary) Synthesis
=========================
```

ajavadia commented on qiskit/synthesis/__init__.py:
```diff
@@ -79,6 +79,13 @@
    SolovayKitaevDecomposition
    generate_basic_approximations
 
+SU(4) (2-qubit unitary) Synthesis
+=========================
```
```suggestion
SU(4) (2-qubit unitary) Synthesis
=================================
```

Cryoris commented on qiskit/circuit/library/standard_gates/siswap.py:
```diff
@@ -0,0 +1,77 @@
+                 └────────────┘└────────────┘
+
+    .. math::
+        B\ q_0, q_1 =
```
Should this be 
```suggestion
        \sqrt{\mathrm{iSwap}}\ q_0, q_1 =
```
?

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+
+    Reference:
+        1. C. Huang et al,
+           *Towards ultra-high fidelity quantum operations:
+           SQiSW gate as a native two-qubit gate (2021)*
```
A bit nitpicky, but I don't think we have slanted fonts anywhere else 🙂 
```suggestion
           Towards ultra-high fidelity quantum operations:
           SQiSW gate as a native two-qubit gate (2021)
```

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+        """Decompose a two-qubit unitary into using the sqrt(iSWAP) gate.
+
+        Args:
+            unitary: a 4x4 unitary to synthesize.
+            basis_fidelity: Fidelity of the iSWAP gate.
+            approximate: Approximates if basis fidelities are less than 1.0.
+            _num_basis_uses: force a particular approximation by passing a number in [0, 3].
```
```suggestion
            unitary: A 4x4 unitary to synthesize.
            basis_fidelity: Fidelity of the iSWAP gate.
            approximate: Approximates if basis fidelities are less than 1.0.
            _num_basis_uses: Force a particular approximation by passing a number in [0, 3].
```

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+        Returns:
+            QuantumCircuit: Synthesized circuit.
+        """
+        if not approximate:
```
That's a more general question: isn't `basis_fidelity` and `approximate` redundant?

ajavadia replied:
hmm good question. I was copying the structure of the XXDecomposer code. I can think about removing `approximate` but I have to see if there is some underlying reason. I'm making some changes to the whole qiskit.synthesis folder so I can have this on the list.

jlapeyre replied:
`approximate` is used by `XXDecomposer` only when computing infidelities. The method used doesn't depend on the values of the fidelities in `basis_fidelity`, in particular does not depend on whether they are less than one. I don't know "less than one" means, maybe it's advice on what to choose if you know you are passing 1.0. The same holds if `basis_fidelity` is not a `dict`, but a float, as it is here. In this case infidelities are computed for several values of `strength` (which corresponds to a rotation angle I think) that are hardcoded in `XXDecomposer`. The infidelities are computed from the strengths by a linear model. If `approximate` is `True` the parameters of the linear model are determined from `basis_fidelity`. If `approximate` is `False` then `basis_fidelity` is ignored completely and infidelities are computed without any reference to parameters passed when calling the object nor those passed when constructing it. That's sort of the reverse of what I would expect, but that's what the code does.

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+            circuit = circuit.compose(red_decomp, [0, 1])
+            if eigenphase_crossing:
+                # y, z -> -z, -y
+                circuit = circuit.compose(XGate().power(1 / 2), [0], front=True)
```
Is there a reason to not use the `SX(dg)Gate` for this? 🙂

ajavadia replied:
hmm no its the same

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+    if abs(C) < _EPS:
+        C = 0.0
+
+    α = np.arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) + 2 * np.sqrt(C))
```
I don't know what the best practice for non-ASCII characters is -- my feeling would say to rather use `alpha` etc to make life easier for future developers 😉 but I'm not fixed on this if you want to keep it as is 🙂

ajavadia replied:
it's just nicer i think to read :)

jlapeyre replied:
I really like reading $\alpha$ in code. But I really don't enjoy searching for it with my editor.

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+    larger volume when approximation is enabled.
+
+    Args:
+        euler_basis (list(str)): single-qubit gates basis in the decomposition.
```
What do you think about adding an example on how to use this class here? 🙂 Would be nice to advertise this feature to users.

ajavadia replied:
I'll add.

Cryoris commented on releasenotes/notes/sqisw-decomposition-2478b9f0b7c54044.yaml:
```diff
@@ -0,0 +1,12 @@
+
+    An associated decomposition method that synthesizes arbitrary two-qubit
+    unitaries into two or three applications of SiSwap has been added in
+    :class:`.SiSwapDecomposer`.
```
It would be cool to show an example in the release notes, too 😄

ajavadia replied:
I'll add.

Cryoris commented on test/python/synthesis/test_siswap_synthesis.py:
```diff
@@ -0,0 +1,137 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
```
is this only because of the variable `u`? If yes maybe we could just rename that one and avoid a global disable 🙂

ajavadia replied:
yes. these variable naming rules are pretty annoying I think. I think we need to relax this at the qiskit level. Like why should naming my unitary `u` be a problem? @mtreinish what do you think?

jakelishman replied:
I'm in favour of relaxing the lint-enforced naming rules in general, although I've got to say the code in this PR isn't exactly the greatest advert for it.  There's a _lot_ of mix-and-match of variable styles with no clear relation to the mathematical paper they're drawn from (the non-iSWAP-specific code), and "why should naming my unitary `u` be a problem?" is indicative - `u` is mostly a bad name, because the majority of things we deal with in Qiskit are unitary.  In tests you can use some name that identifies it as the base of the test, and in synthesis code, I'm guessing there's _a lot_ of different unitaries kicking around at different points, so they ought to have proper names as well.

jlapeyre replied:
But often the main point of a function is to do something with a single unitary. In this case naming it `unitary` seems reasonable. `XXDecomposer` does this and it looks pretty good there. I don't like `u`, but if I used a editor that had good support for Python semantics, I might not mind it. I have to distinguish occurrences of `u` visually. If it's really just as easy with a modern editor to deal with, I'd say go ahead and use it.

Cryoris commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+
+class SiSwapDecomposer:
+    """
+    A class for decomposing 2-qubit unitaries into at most 3 uses of the sqrt(iSWAP) gate.
```
In the PR text you're mentioning this code differs from the paper, would it make sense to summarize the differences either here or as comments in the code to know what's going on when we come back to this code in the future?

ajavadia replied:
I'll add.

Cryoris reviewed (commented):
Thanks for the updates and the additional tests! I left some more comments below. 🙂

jakelishman commented on qiskit/circuit/library/standard_gates/siswap.py:
```diff
@@ -0,0 +1,77 @@
+        self.definition = qc
+
+    def __array__(self, dtype=None):
+        """Return a numpy.array for the DCX gate."""
```
```suggestion
        """Return a numpy.array for the :math:`\\sqrt{iSWAP}` gate."""
```

perhaps?

jakelishman commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+           `arXiv:2105.06074 <https://arxiv.org/abs/2105.06074>`_
+    """
+
+    def __init__(self, euler_basis: Optional[List[str]]):
+        # decomposer for the local single-qubit gates
+        from qiskit.transpiler.passes.optimization.optimize_1q_decomposition import (
+            Optimize1qGatesDecomposition,  # pylint: disable=cyclic-import
+        )
+
+        euler_basis = euler_basis or ["u"]
+        self._decomposer1q = Optimize1qGatesDecomposition(euler_basis)
```
I don't know how much you mind about the type hints, but `List` is only _very_ rarely want to meant to accept when it's an input.  Most likely, you just want anything that can be converted to a list, which is `typing.Iterable[str]`.  If so, you'd need to change your assignment code to `euler_basis = ["u"] if euler_basis is None else list(euler_basis)` to be properly correct.

It also _probably_ doesn't matter, but your current code will convert `euler_basis=[]` to `["u"]`.  I don't know if that's intended, or if it's something you'd want `Optimize1qGatesDecomposition` to throw an error on.  I would also mention that the more ideal case would be making a dummy `Target` than passing the basis directly, because Matthew wants to start getting all transpiler passes transitioned to remove the old-style interfaces for easier maintainability moving forwards.

jakelishman commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+            circuit = circuit.compose(XGate(), [0], front=True)
+            circuit = circuit.compose(ZGate(), [1], front=True)
+            circuit = circuit.compose(YGate(), [1])
+    else:
+        if z < 0:  # blue region
```
The body of this `else` is just an `if/else` itself - you could make line 163 an `elif` and remove a level of indentation.

jakelishman commented on qiskit/synthesis/su4/siswap_decompose.py:
```diff
@@ -0,0 +1,286 @@
+    (x, y, z) ∈ W′ when sandwiched by two SiSwap gates.
+    Return the SiSwap sandwich.
+    """
+    C = np.sin(x + y - z) * np.sin(x - y + z) * np.sin(-x - y - z) * np.sin(-x + y + z)
+    if abs(C) < _EPS:
+        C = 0.0
+
+    α = np.arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) + 2 * np.sqrt(C))
```
Wherever you're dealing with scalars and not things that might be Numpy arrays to be vectorised over, it's better to use the functions from `math` or `cmath` as appropriate rather than `numpy`, because `numpy` ones have a really very sizeable dispatch overheard when used on scalars.

jlapeyre replied:
I knew it was there, but I'm surprised how large it is. I just did a test with `npsin = np.sin` and `mathsin = math.sin`. timeit showed numpy more than 10x slower.

jakelishman commented on qiskit/synthesis/su4/utils.py:
```diff
@@ -0,0 +1,88 @@
+    if len(P) == 1:
+        return P[0]
+
+    P = [np.array(p) for p in P]
```
You probably mean `np.asarray` (everywhere in this function), since you're not trying to mutate things or need contiguous memory, and you probably don't need a copy made if things are already an array.

jakelishman commented on qiskit/synthesis/su4/utils.py:
```diff
@@ -0,0 +1,88 @@
+
+    P is given as a list of points forming its vertices.
+
+    Function from
+    https://scipy-cookbook.readthedocs.io/items/Finding_Convex_Hull_Minimum_Point.html
```
It's not clear to me what the licence of this code is.  The best I can find is in the repo of the docs website, which has a stipulation that source-code reproductions must include their licence header and binary distributions must include it in their documentation.  The licence comes from the times when Scipy was part of Enthought.

I think we need to be more careful about just copy-pasting code like this - it's not clear to me if we're actually safe to use this.

jakelishman commented on qiskit/synthesis/su4/utils.py:
```diff
@@ -0,0 +1,88 @@
+    a0, b0, c0 = p
+    a1, b1, c1 = q
+
+    return 1 - 1 / 20 * (
+        4
+        + 16
+        * (
+            np.cos(a0 - a1) ** 2 * np.cos(b0 - b1) ** 2 * np.cos(c0 - c1) ** 2
+            + np.sin(a0 - a1) ** 2 * np.sin(b0 - b1) ** 2 * np.sin(c0 - c1) ** 2
+        )
+    )
```
Similar comment about preferring `math` functions if appropriate here.

jakelishman commented on test/python/synthesis/test_siswap_synthesis.py:
```diff
@@ -0,0 +1,137 @@
+    @combine(seed=range(50))
+    def test_siswap_random(self, seed):
+        """Test synthesis of 50 random SU(4)s."""
+        u = random_unitary(4, seed=seed)
+        decomposer = SiSwapDecomposer(euler_basis=["rz", "ry"])
+        circuit = decomposer(u)
+        self.assertLessEqual(circuit.count_ops().get("siswap", 0), 3)
+        self.assertEqual(Operator(circuit), Operator(u))
```
In all these in this file, should we also be testing that the only gates in the output are those in the Euler basis as well?

jakelishman commented on test/python/synthesis/test_siswap_synthesis.py:
```diff
@@ -0,0 +1,137 @@
+        self.assertEqual(circuit.count_ops().get("siswap", 0), 0)
+        self.assertEqual(Operator(circuit), Operator(u))
+
+    @combine(
+        p=[
+            [np.pi / 4, np.pi / 6, -np.pi / 12],  # red and blue intersection (bottom face)
+            [np.pi / 4, np.pi / 6, np.pi / 12],  # inverse of above (and locally equivalent)
+            [np.pi / 4, np.pi / 8, np.pi / 8],  # half-way between CX and SWAP (peak red polytope)
+            [np.pi / 4, np.pi / 8, -np.pi / 8],  # inverse of above (and locally equivalent)
+            [np.pi / 4, np.pi / 16, np.pi / 16],  # quarter-way between CX and SWAP
+            [np.pi / 4, np.pi / 16, np.pi / 16],  # inverse of above (and locally equivalent)
+            [np.pi / 6, np.pi / 8, np.pi / 24],  # red and blue and green intersection
+            [np.pi / 6, np.pi / 8, -np.pi / 24],  # inverse of above (not locally equivalent)
+            [np.pi / 16, np.pi / 24, np.pi / 48],  # red and blue and purple intersection
+            [np.pi / 16, np.pi / 24, -np.pi / 48],  # inverse of above (not locally equivalent)
+            [np.pi / 6, np.pi / 12, -np.pi / 16],  # inside red polytope
+        ]
+    )
+    def test_siswap_special_points(self, p):
```
It's really good to see this test - also great to have very explicit testing of a load of points we know to be special.

ShellyGarion commented on qiskit/circuit/library/__init__.py:
```diff
@@ -113,6 +113,7 @@
    SdgGate
    SwapGate
    iSwapGate
+   SiSwapGate
```
The SiSwap Gate should also be added to the documentation in this init file, and removed from the docs in circuit/library/standard_gates. 
See the following PR: #9668

ShellyGarion commented on qiskit/circuit/library/standard_gates/__init__.py:
```diff
@@ -11,7 +11,67 @@
 
 """
-Standard gates
+=============================================================
+Standard gates (:mod:`qiskit.circuit.library.standard_gates`)
+=============================================================
```
Indeed, see the details in PR #9668. Please add this gate to the docs in the init file of circuit/library.

mtreinish commented:
Is there any update on this? Is this something we want to target for 0.45.0?

evmckinney9 commented:
> Is there any update on this? Is this something we want to target for 0.45.0?

Hello, hoping this feature gets merged eventually. 
Is there anything I might be able to do to help finalize this PR?
For reference, [here is a notebook](https://gist.github.com/evmckinney9/30845d91d3e85cb6bdf962fb090abc37) I use for compiling into the sqrtiswap basis with this PR that I have moved into a fork since I find this functionality very necessary.
